### PR TITLE
Refactor data layer and fix request loan success details not showing

### PR DESCRIPTION
**Pull request by chrismin**

### CODE REVIEW

File: src/models/DebtOrders/DebtOrder.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+     *
+     * https://stackoverflow.com/questions/45964008/typescript-instanceof-not-working
+     */
+    constructor(params?: any) {
+        if (params) {
+            for (const key of Object.keys(params)) {
+                this[key] = params[key];
+            }
+        }
+    }
```
Comment (nadavhollander) [inline src/models/DebtOrders/DebtOrder.ts]: Does a constructor-initialized type in TS still have the same strictness in terms of type checking as a normal typescript object? i.e.if a user leaves out `amortizationUnit` in`params` will typescript flag it in the IDE?

File: src/models/DebtOrders/FilledCollateralizedDebtOrder.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import { BigNumber } from "bignumber.js";
+import { FilledDebtOrder } from "./FilledDebtOrder";
+import { CollateralParameters } from "../DebtInterfaces";
+
+export class FilledCollateralizedDebtOrder extends FilledDebtOrder implements CollateralParameters {
+    collateralAmount: BigNumber; // raw amount
+    collateralTokenSymbol: string;
+    gracePeriodInDays: BigNumber;
+    collateralReturnable: boolean;
+    // collateralSeizable: boolean;
```
Comment (nadavhollander) [inline src/models/DebtOrders/FilledCollateralizedDebtOrder.ts]: Why is this commented out?

File: src/models/DebtQueryParams.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    underwriterRiskRating?: BigNumber;
+    termsContract?: string;
+    termsContractParameters?: string;
+    expirationTimestampInSec?: BigNumber;
+    salt?: BigNumber;
+
+    // Signatures
+    debtorSignature?: object;
+    creditorSignature?: object;
+    underwriterSignature?: object;
```
Comment (nadavhollander) [inline src/models/DebtQueryParams.ts]: This looks identical to the `DebtOrder.Instance` object -- why are we not using that instead?

File: src/modules/FillLoan/FillLoanEntered/FillLoanEntered.tsx [hunk 4]
```diff
@@ -47,7 +47,8 @@ interface States {
     collateralized?: boolean;
     collateralTokenSymbol?: string;
     confirmationModal: boolean;
-    debtOrder: DebtOrder.Instance;
+    debtOrder?: OpenCollateralizedDebtOrder;
+    debtOrderInstance?: DharmaDebtOrder.Instance;
```
Comment (nadavhollander) [inline src/modules/FillLoan/FillLoanEntered/FillLoanEntered.tsx]: This seems redundant -- why is this structured like so?

File: __test__/unit/modules/Dashboard/Debts/ActiveDebtOrder/ActiveDebtOrder.test.tsx [hunk 5]
```diff
@@ -39,30 +39,41 @@ import { BigNumber } from "bignumber.js";
 import MockDharma from "__mocks__/dharma.js";
 import { TokenAmount } from "src/components";
 import { ScheduleIcon } from "src/components/scheduleIcon/scheduleIcon";
+import { DebtOrder as DharmaDebtOrder } from "@dharmaprotocol/dharma.js/dist/types/src/types";
```
[the remark this answers is not in the document]
Comment (chrismin) [reply __test__/unit/modules/Dashboard/Debts/ActiveDebtOrder/ActiveDebtOrder.test.tsx]: agreed!

File: __test__/unit/modules/Dashboard/Debts/DebtsMetrics/DebtsMetrics.test.tsx [hunk 6]
```diff
@@ -235,6 +126,106 @@ describe("<DebtsMetrics />", () => {
...
+                    issuanceHash:
+                        "0x89e9eac37c5f14b657c69ccd891704b3236b84b9ca1d449bd09c5fbaa24afebf",
+                    repaidAmount: new BigNumber(4),
+                    repaymentSchedule: [1553557371],
+                    creditor: "0x431194c3e0f35bc7f1266ec6bb85e0c5ec554935",
+                    description: "Hello, Can I borrow some OMG please?",
+                    fillLoanShortUrl: "http://bit.ly/2I4bahM",
+                    totalExpectedRepayment: new BigNumber(
+                        await dharma.servicing.getTotalExpectedRepayment(),
+                    ),
```
[the remark this answers is not in the document]
Comment (chrismin) [reply __test__/unit/modules/Dashboard/Debts/DebtsMetrics/DebtsMetrics.test.tsx]: Done.

File: src/models/DebtOrders/DebtOrder.ts (hunk 1 above)
Comment (chrismin) [reply src/models/DebtOrders/DebtOrder.ts]: Unfortunately neither the IDE nor the compiler seems to catch that... I will look into alternatives and workarounds...

The ultimate alternative is to refactor the classes into interfaces, but then the "typeguarding" gets a little messier :(

File: src/models/DebtOrders/FilledCollateralizedDebtOrder.ts (hunk 2 above)
Comment (chrismin) [reply src/models/DebtOrders/FilledCollateralizedDebtOrder.ts]: Done.

File: src/models/DebtQueryParams.ts (hunk 3 above)
Comment (chrismin) [reply src/models/DebtQueryParams.ts]: The class extends OpenCollateralizedDebtOrder, as well as contains all of DebtOrder.Instance's properties.  

Unfortunately, ts doesn't support multiple inheritance (which would have been useful and cleaner!).

Another consideration is that we don't have direct/clean access to dharma.js's ECDSASignature class, which means instatiating such objects on the Plex side is tricky, especially when parsing from raw JSON in something like the Fill Loan component.

File: src/models/DebtInterfaces/CollateralParameters.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+import { BigNumber } from "bignumber.js";
+
+export interface CollateralParameters {
+    collateralAmount: BigNumber; // raw amount
```
Comment (saturnial) [inline src/models/DebtInterfaces/CollateralParameters.ts]: Why isn't this a `TokenAmount`?

File: src/models/DebtOrders/DebtOrder.ts (hunk 1 above)
Comment (saturnial) [reply src/models/DebtOrders/DebtOrder.ts]: why can't this be an interface? because of union types?

File: src/models/DebtOrders/OpenDebtOrder.ts [hunk 8]
```diff
@@ -0,0 +1,3 @@
+import { DebtOrder } from "./DebtOrder";
+
+export class OpenDebtOrder extends DebtOrder {}
```
Comment (saturnial) [inline src/models/DebtOrders/OpenDebtOrder.ts]: There is nothing that uniquely describes an open debt order?

File: src/models/DebtParameters/DebtParameters.ts [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import { BigNumber } from "bignumber.js";
+// import { AmmortizationUnit } from "@dharmaprotocol/dharma.js/dist/lib/src/adapters/simple_interest_loan_adapter";
+
+export class DebtParameters {
+    amortizationUnit: any; // TODO: export AmmortizationUnit from dharma.js
```
Comment (saturnial) [inline src/models/DebtParameters/DebtParameters.ts]: lol. any? really?

File: src/models/DebtInterfaces/CollateralParameters.ts (hunk 7 above)
Comment (chrismin) [reply src/models/DebtInterfaces/CollateralParameters.ts]: `TokenAmount` returns a JSX element, which we can't use as a value in a class or interface (or, rather, in this case we wouldn't want to...)

What we really need here is the TokenAmount class from dharma.js!!

File: src/models/DebtParameters/DebtParameters.ts (hunk 9 above)
Comment (chrismin) [reply src/models/DebtParameters/DebtParameters.ts]: err.... sorry haha 😅  will remove in a separate PR

**Pull request merged**